src/Units/KaijuActions.java

```diff
@@ -0,0 +1,943 @@
+   * <p>Designed for Kaiju ability actions
+   */
+  public static void enqueueKaijuStrikeEvents(MapMaster gameMap, Unit kaiju, XYCoord xyc, GameEventQueue events)
+  {
+    if( enqueueKaijuKillEvents(gameMap, kaiju, xyc, events) )
+      return; // Kaiju resistance cancels all events
+
+    MapLocation location = gameMap.getLocation(xyc);
+
+    if( isCrushable(location) )
+    {
+      Environment oldEnvirons = location.getEnvironment();
+      Environment newEnvirons = Environment.getTile(oldEnvirons.terrainType.getBaseTerrain(), oldEnvirons.weatherType);
+      events.add(new MapChangeEvent(location.getCoordinates(), newEnvirons));
+
+      if( Utils.willLoseFromLossOf(gameMap, location) )
+      {
+        events.add(new ArmyDefeatEvent(location.getOwner().army));
+      }
+    }
+  }
+  /**
+   * Destroy a non-Kaiju-resistor on this spot
+   * @return Whether there was a Kaiju-resistor
+   */
+  public static boolean enqueueKaijuKillEvents(MapMaster gameMap, Unit kaiju, XYCoord xyc, GameEventQueue events)
+  {
+    MapLocation location = gameMap.getLocation(xyc);
+
+    Unit victim = location.getResident();
+    if( null != victim )
+    {
+      KaijuWarsUnitModel kjum = (KaijuWarsUnitModel) victim.model;
+      if( kjum.resistsKaiju )
+        return true; // Kaiju resistance cancels all events
+      else
+      {
+        ArrayList<Unit> stompable = new ArrayList<>();
+        stompable.add(victim);
+        events.add(new MassDamageEvent(kaiju.CO, stompable, victim.getHP(), true));
+        Utils.enqueueDeathEvent(victim, events);
+      }
+    }
+
+    return false;
+  }
+
+  /** Superclass for Kaiju attack ability actions */
+  public abstract static class KaijuAttackAction extends GameAction
+  {
+    final UnitActionFactory type;
+    final Class<?> abilityKey;
+    final Unit actor;
+    final XYCoord target;
+    final boolean destroyBuildings;
+    boolean longCooldown = false;
+
+    public KaijuAttackAction(UnitActionFactory pType, Class<?> pKey, Unit unit, XYCoord target, boolean destroyBuildings)
+    {
+      type = pType;
+      abilityKey = pKey;
+      actor = unit;
+      this.target = target;
+      this.destroyBuildings = destroyBuildings;
+    }
+
+    @Override
+    public GameEventQueue getEvents(MapMaster gameMap)
+    {
+      GameEventQueue eventSequence = new GameEventQueue();
+      if( destroyBuildings )
+        for( XYCoord xyc : getTargets(gameMap) )
+          enqueueKaijuStrikeEvents(gameMap, actor, xyc, eventSequence);
+      else
+        for( XYCoord xyc : getTargets(gameMap) )
+          enqueueKaijuKillEvents(gameMap, actor, xyc, eventSequence);
+
+      KaijuStateTracker kaijuTracker = StateTracker.instance(gameMap.game, KaijuStateTracker.class);
+      // Setting the tracker state here feels wrong
+      if( longCooldown )
+        kaijuTracker.abilityUsedLong(actor, abilityKey);
+      else
+        kaijuTracker.abilityUsedShort(actor, abilityKey);
+
+      return eventSequence;
+    }
```

Comment:
Tracker information is properly part of the game state, I think... which makes setting that in `getEvents()` feel a little weird.
On the other hand, this *works* just fine since `getEvents()` is only ever called once per action and has explicit `MapMaster` access.
Also worth noting is that we do a lot of similar side-effect-y stuff in event listener hooks.

That said, I feel like maybe all changes to the game state proper should go through events, so nothing outside of explicit mutators has side effects.
This would have the nice property that you can reproduce any game state with the start and a stream of events - no listeners required at all.

There are two approaches I can see to achieve the above dream:
Add a bunch of events, one for each potential gamestate mutation
Add a generic event you can throw "whatever" in, i.e. my `LambdaEvent` from a while back (we could throw in a `name` field or something for easier debugging?)

Another reasonable approach would be to just accept that event listen hooks (and possibly `getEvents()` itself) can have side effects on the game state.
I would probably want some nomenclature change to make that more clear.
There is one notable drawback over the above two possibilities, as well: you don't get to order your changes arbitrarily - if I do something that I don't have an event for, it must happen before any events are executed.

Reply:
> Tracker information is properly part of the game state, I think... which makes setting that in getEvents() feel a little weird.

Yeah - is there any reason you can't set the cooldown when the events are executed?

> That said, I feel like maybe all changes to the game state proper should go through events, so nothing outside of explicit mutators has side effects.

That is the direction I was trying to take things when events were initially introduced.

> Add a bunch of events, one for each potential gamestate mutation

I mean, that's basically what we do, right? There are a couple of combo ones, like I think the unit-healing event also handles the payment portion and maybe shouldn't.

> generic event

I think I'd lean against this by default. Not sure what would necessitate events with unknown effects.

> Another reasonable approach would be to just accept that event listen hooks (and possibly getEvents() itself) can have side effects on the game state.

Event listen hooks (by design) can inject new events into the stream - this happens after the listened event is executed though. Granted, a CO could potentially change internal state without an event based on receiving one, but I have less heartburn about that than about using `getEvents` to generate side effects.

Reply:
> Yeah - is there any reason you can't set the cooldown when the events are executed?

There's no event for it, and no unique event that the cooldown tracker could listen to.
> That is the direction I was trying to take things when events were initially introduced.

Should that also include event listeners publishing their own events to make changes to the gamestate?
> I mean, that's basically what we do, right? There are a couple of combo ones, like I think the unit-healing event also handles the payment portion and maybe shouldn't.

Yeah. There's so much boilerplate, though.
> I think I'd lean against this by default. Not sure what would necessitate events with unknown effects.

Nothing *necessitates* it. But it'd be really nice to have a facility for doing one-off "change this number in a `StateTracker`" stuff without writing a whole new class. If it'd make you feel better, we could require a description for the `ToString()` for `LambdaEvent` so it's easier to debug.
If we didn't have this and did move to "listeners may only make changes by publishing more events", then every `StateTracker` would have to define at least one new type of event.
> Event listen hooks (by design) can inject new events into the stream - this happens after the listened event is executed though.

Two fine points here:
They initially could not inject new events, so a number of listeners were written without that assumption.
The event is executed before it is published - a listener "direct effect" takes place *immediately after* the event, while injected events will go to the end of the *event sequence* from the last `GameAction` taken.
> Granted, a CO could potentially change internal state without an event based on receiving one, but I have less heartburn about that than about using `getEvents` to generate side effects.

Yeah. I would still like to do something with nomenclature to make it clearer that that's kosher, though (or decide it isn't kosher).

Reply:
> Should that also include event listeners publishing their own events to make changes to the gamestate

As in, do we need  meta-event for "I just created more events"? Probs not. Do you just mean "Should publishing events on receive be allowed"? Pretty sure we already do that.

> a listener "direct effect" takes place immediately after the event, while injected events will go to the end of the event sequence from the last GameAction taken.

Is that a problem? I guess you could just immediately execute/publish a sub-event instead of simply queuing it if it makes a big difference.

Reply:
> As in, do we need meta-event for "I just created more events"? Probs not. Do you just mean "Should publishing events on receive be allowed"? Pretty sure we already do that.

That's not where I was going with that.
To rephrase: Should event listeners be able to make changes to the gamestate _without_ publishing an event?
Current usecases of that are COs and all StateTrackers.
The former probably should move to publishing events for ability charge.
The latter probably shouldn't have to make a new type of event for any given change, which implies either "allow it" or "add a generic event".
> > a listener "direct effect" takes place immediately after the event, while injected events will go to the end of the event sequence from the last GameAction taken.
> 
> Is that a problem? I guess you could just immediately execute/publish a sub-event instead of simply queuing it if it makes a big difference.

It may or may not be a problem in the future. It isn't one right now. I just wanted to clarify.